### PR TITLE
Replace `sha1` npm package with inline implementation

### LINKED ISSUE #1464: snowplows `sha1` dependency pulls 28 KB Buffer polyfill into browser bundles
`@snowplow/browser-tracker-core` depends on the 11y old [`sha1`](https://www.npmjs.com/package/sha1) npm package (v1.1.1) and that `sha1` package uses `Buffer.isBuffer()` internally to type-check its input:

```js
// node_modules/sha1/sha1.js
else if (typeof Buffer !== 'undefined' && typeof Buffer.isBuffer == 'function' && Buffer.isBuffer(message))
```

Snowplow only ever passes a string to `sha1` (cookie domain + path for the domain hash), so this `Buffer` code path is never reached at runtime. But compilers (e.g. webpack) doesn't know that at build time

Next.js ships a `ProvidePlugin` that auto-injects `Buffer` from `next/dist/compiled/buffer/index.js` whenever it sees a bare `Buffer` reference in any module. Since `sha1` references `Buffer`, the entire **~28 KB Buffer polyfill** gets bundled into every client-side page even though it's dead code

This affects the global bundle (`_app` chunk), so every single page pays the cost of all next.js apps which use snowplow

Also the published ESM dist (`dist/index.module.js`) does `import hash from 'sha1'`, but `sha1` itself only ships CommonJS (`module.exports = api`).. 

That's why snowplow forces all bundlers to activate their CJS interop mode which can also interfere with tree-shaking depending on the bundler

```
@snowplow/browser-tracker-core (ESM)
  → sha1 (CJS, references Buffer)
    → next/dist/compiled/buffer/index.js (+28 KB polyfill)
```

The only usage is in `libraries/browser-tracker-core/src/tracker/index.ts`:

```ts
domainHash = hash((configCookieDomain || domainAlias) + (configCookiePath || '/')).slice(0, 4);
```

## Impact

- +28 KB (raw) / ~7 KB (gzipped) added to every page's initial load
- Affects all Next.js / webpack 5 users of `@snowplow/browser-tracker`
- The `sha1` package hasn't been updated in years and also depends on `charenc` and `crypt`
- Workaround: manually strip `Buffer` from webpack's `ProvidePlugin` and `resolve.fallback` in next.config — brittle and easy to regress

Image

## Environment

- `@snowplow/browser-tracker-core` 4.6.9
- Next.js 16.1.7 (webpack 5)

### PR DESCRIPTION
**Pull request by jantimon**
Fixes #1464

The `sha1` npm package references `Buffer.isBuffer()` internally. Bundlers like webpack/Next.js see this and inject a **~28 KB Buffer polyfill** into every client page — even though the `Buffer` code path is never reached (we only hash short strings)

On top of that, `sha1` only ships CommonJS, forcing CJS interop on every bundler that consumes the ESM dist

- Remove `sha1` and `@types/sha1` dependencies
- Add a ~90-line inline SHA-1 (FIPS 180-4) that only handles UTF-8 strings — which is all the tracker needs
- Tests cover RFC 3174 vectors plus the actual domain-hash values to ensure identical output

- **-28 KB raw / ~7 KB gzipped** from every page for Next.js / webpack 5 users
- Eliminates the last CJS dependency in `browser-tracker-core`
- No API or behavioral changes

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

**Comment by snowplowcla** [comment]
Confirmed! @jantimon has signed the Contributor License Agreement. Thanks so much.

File: libraries/browser-tracker-core/src/helpers/sha1.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  for (let i = 0; i < message.length; i++) {
+    let c = message.charCodeAt(i);
+    if (c < 0x80) {
+      bytes.push(c);
+    } else if (c < 0x800) {
+      bytes.push(0xc0 | (c >> 6), 0x80 | (c & 0x3f));
+    } else if (c >= 0xd800 && c <= 0xdbff && i + 1 < message.length) {
+      const next = message.charCodeAt(++i);
+      c = 0x10000 + ((c & 0x3ff) << 10) + (next & 0x3ff);
+      bytes.push(0xf0 | (c >> 18), 0x80 | ((c >> 12) & 0x3f), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
```
[the remark this answers is not in the document]
Comment (jantimon) [reply libraries/browser-tracker-core/src/helpers/sha1.ts:19]: if you take a look at the original code 
https://github.com/pvorb/node-charenc/blob/c92eade4dcb3d5c43a14f7c45e7af7dee562723e/charenc.js#L4-L7

the original sha1 package's UTF-8 encoding just delegates to encodeURIComponent under the hood. 

`encodeURIComponent` doesn't emit `U+FFFD` for malformed surrogates it straight up throws a URIError. so the old package would crash on ill-formed input

my implementation is actually more lenient in that regard not less

applying this copilot suggestion would actually introduce a regression

File: libraries/browser-tracker-core/src/helpers/sha1.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    } else {
+      bytes.push(0xe0 | (c >> 12), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
+    }
+  }
+
+  // Pre-processing: pad to 512-bit blocks
+  const bitLen = bytes.length * 8;
+  bytes.push(0x80);
+  while (bytes.length % 64 !== 56) bytes.push(0);
+  // Append length as 64-bit big-endian (high 32 bits always 0 for strings < 512 MB)
```
[the remark this answers is not in the document]
Comment (jantimon) [reply libraries/browser-tracker-core/src/helpers/sha1.ts:29]: the original implementation does `m[((l + 64 >>> 9) << 4) + 15] = l` where `l = message.length * 8`:

https://github.com/pvorb/node-sha1/blob/0843eb6a4a4192473deb0fc3360ab2d50c7d9b39/sha1.js#L29

that's a 32-bit integer assignment 
 
 so it silently wraps for large inputs without any guard or error

my implementation has the same effective cap and is not more broken in that edge case than the original

also I am not sure when snowplow would have hashes for 1/2 GB of data in the browser

File: common/autoinstallers/rush-prettier/pnpm-lock.yaml [hunk 3]
```diff
@@ -1,296 +1,263 @@
...
+importers:
+
+  .:
+    dependencies:
+      prettier:
+        specifier: ^2.5.1
+        version: 2.8.8
+      pretty-quick:
+        specifier: ^3.1.3
+        version: 3.3.1(prettier@2.8.8)
```
[the remark this answers is not in the document]
Comment (jantimon) [reply common/autoinstallers/rush-prettier/pnpm-lock.yaml]: this is a side effect of running rush update on the autoinstaller - not sure why this is not on main - should I revert this change?

File: libraries/browser-tracker-core/test/helpers/sha1.test.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    expect(sha1('')).toBe('da39a3ee5e6b4b0d3255bfef95601890afd80709');
+  });
+
+  test('"abc"', () => {
+    expect(sha1('abc')).toBe('a9993e364706816aba3e25717850c26c9cd0d89d');
+  });
+
+  test('"The quick brown fox jumps over the lazy dog"', () => {
+    expect(sha1('The quick brown fox jumps over the lazy dog')).toBe('2fd4e1c67a2d28fced849ee1bb76e7391b93eb12');
+  });
```
[the remark this answers is not in the document]
Comment (jantimon) [reply libraries/browser-tracker-core/test/helpers/sha1.test.ts:15]: the original package's own test suite is 3 tests: one ASCII string, one probabilistic collision check (which doesn't verify any known vector), and one Buffer test that's not even applicable here:

https://github.com/pvorb/node-sha1/blob/master/test.js

so the mentioned edge cases by copilot aren't tested there as well

but i am happy to add non-ASCII vectors if you think that's helpful

File: common/autoinstallers/rush-prettier/pnpm-lock.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jantimon) [reply common/autoinstallers/rush-prettier/pnpm-lock.yaml]: reverted

**Comment by jantimon** [comment]
@matus-tomlein all @copilot remarks have been resolved - could you please take a look?

**Review by matus-tomlein — APPROVED** [review]
Thank you for the contribution @jantimon, that's a great improvement in the package size! Have run some more tests and everything seems to be correct, so we will try to get this out in a release soon.

**Comment by matus-tomlein** [comment]
@jantimon could you please point this PR into the `release/4.7.0` branch?

**Comment by jantimon** [comment]
@matus-tomlein done - had to fix the merge conflicts but it looks good now
